Prevent users from adding permissions they don't have to a group

TW80000 opened the pull request:
- Anyone can create or update a group if they have permission (normal AuthenticatedDAO already exists on groupDAO), but this PR adds an extra check to make sure you can't add any permissions that aren't implied by ones you already have to an existing group or a newly created group.

Marks https://github.com/nanoPayinc/NANOPAY/issues/3130 ready for testing.

Commit:
Prevent users from updating their own group

Still allow administrators to set the group of another user.

kgrgreer commented:
I think this needs to be more sophisticated than this. Let's discuss in person tomorrow.

TW80000 commented:
@kgrgreer Said the following in Slack on Monday, August 13th, 2018:

> Any user should be able to set the group of any user whose group they have update permission to, to any other group for whom they also have update permission to. They should also be able to update those groups, but not to add any permission which they don't themselves have. This prevents them from ever escalating their permissions. This would effectively accomplish the same thing as not enabling them to change their group. They could change their own group, but only ever down-grade or side-grade it, never upgrade.

We never discussed it in person, but I'll assume this comment covers the requirements.

kgrgreer commented on src/foam/nanos/auth/AuthenticatedUserDAO.java:
```diff
@@ -67,6 +72,21 @@ public FObject put_(X x, FObject obj) {
+        if ( updatingSelf ) {
+          throw new RuntimeException("You cannot set your own group.");
+        } else if ( ! isGlobalAdmin ) {
+          throw new RuntimeException("Only administrators may update another user's group.");
```
I don't think this message or code is entirely accurate. I would just say that you aren't authorized to update the user's group. Also, I don't think we should check if you're a global admin, just if you have access to this group and user. We want to have weaker admins, like a GS Admin who can administer all of the groups and users under their control.

Commit:
Merge branch 'master' into prevent-setting-own-group

Commit:
Improve AuthenticatedUserDAO in various ways

- global permissions properly end in wildcards
- made permission check for updating users more granular by including
the id in the permission
- remove duplicate null check
- use AuthorizationException instead of RuntimeException
- make error messages more accurate
- don't let users change the group of another user to a group unless
they have update permission on that group

Commit:
Remove comment

TW80000 commented on src/foam/nanos/auth/AuthenticatedUserDAO.java:
```diff
@@ -65,6 +74,23 @@ public FObject put_(X x, FObject obj) {
+      result != null &&
+      ! SafetyUtil.equals(result.getGroup(), toPut.getGroup())
+    ) {
+      boolean hasGroupUpdatePermission = auth.check(x, "group.update." + result.getGroup());
```
@kgrgreer Are we sure "group.update.whatever" is the best permission for this purpose? I'd expect that to let me edit a specific group, but not necessarily assign users to it.

kgrgreer replied:
I think that the two roles should be the same. If I can do one then I should be able to do the other.

Commit:
Prevent permissions escalation by editing/creating groups

kgrgreer commented on src/foam/nanos/auth/AuthenticatedGroupDAO.java:
```diff
@@ -0,0 +1,65 @@
+ * Custom authentication for groupDAO
+ *
+ * Features:
+ *   - If creating a group, you can only give it permissions you already have.
```
"permissions implied by those that you already have"

Which is slightly different.

kgrgreer commented on src/foam/nanos/auth/AuthenticatedGroupDAO.java:
```diff
@@ -0,0 +1,65 @@
+    Group toPut = (Group) obj;
+    String id = toPut.getId();
+    Group existing = (Group) getDelegate().find_(x, id);
+    if ( existing != null ) checkUserHasAllPermissionsInGroup(x, existing);
```
Why do we need to check the old Group?

TW80000 replied:
If we didn't you could completely replace all permissions in a group you had permission to update (`"group.update.example"`), but for which you didn't have all the permissions in that group. I don't know if we want to allow that or not. Would we ever have a situation where you had update permission on a group but didn't have one of the permissions in the group? As far as I know that could happen, whether intentionally or not.

kgrgreer replied:
I still don't think this check is needed. If you have update access to a group then you can probably create new users of that group, in which case you could create yourself a new user to gain that access, so we aren't really protecting anything with this.

kgrgreer commented on src/foam/nanos/auth/AuthenticatedGroupDAO.java:
```diff
@@ -0,0 +1,65 @@
+  @Override
+  public FObject remove_(X x, FObject obj) {
+    Group group = (Group) obj;
+    checkUserHasAllPermissionsInGroup(x, group);
```
I don't think we need to check this.

TW80000 replied:
I guess it doesn't make much sense. You'd still need `"group.remove.example"` anyway.

Commit:
Don't require user to have all permissions in a group to delete it

Commit:
Make documentation more accurate

kirkeaton commented on src/foam/nanos/auth/AuthenticatedGroupDAO.java:
```diff
@@ -0,0 +1,57 @@
+    String groupId = toCheck.getId();
+
+    while ( ! SafetyUtil.isEmpty(groupId) ) {
+      Group group = (Group) getDelegate().find_(x, groupId);
```
The first time this loop runs the Group return will be equal to the Group that you passed in correct? It's minor but it would be more optimized if you eliminated the first call to the DAO.

TW80000 replied:
Yeah, on top of that this actually breaks creating a group too because the `find` will always return null for a new group. Fixing now.

kirkeaton reviewed (changes requested):
Please write tests to verify the functionality here.

kirkeaton commented on src/foam/nanos/auth/AuthenticatedUserDAO.java:
```diff
@@ -65,6 +74,23 @@ public FObject put_(X x, FObject obj) {
+        throw new AuthorizationException("You do not have permission to change this user's group.");
+      } else if ( ! hasGroupUpdatePermission ) {
+        throw new AuthorizationException("You do not have permission to change a user's group to '" + toPut.getGroup() + "'.");
+      }
```
I'd rather see this logic in a separate decorator as to not bloat the AuthenticatedUserDAO too much

kgrgreer replied:
But this logic is related to authorization, so I think it would be confusing to have that responsibility split betwen two AuthenticatedUserDAOs.

kirkeaton commented on src/foam/nanos/auth/AuthenticatedGroupDAO.java:
```diff
@@ -0,0 +1,57 @@
+package foam.nanos.auth;
```
Minor comment, add license header.

Commit:
Merge branch 'master' into prevent-setting-own-group

kgrgreer commented on src/foam/nanos/auth/AuthenticatedGroupDAO.java:
```diff
@@ -0,0 +1,57 @@
+  @Override
+  public FObject put_(X x, FObject obj) {
+    Group toPut = (Group) obj;
+    String id = toPut.getId();
```
Do you really care that its a String? Could you just declare it as Object, which would make your code more resilient to change?

TW80000 commented:
I ran into a problem with this scheme while writing tests for it.

Say I have the `group.create` permission so I can create new groups. I create a group with id `"MyNewGroup"` and give it only permissions already implied by ones I have. All is well. I also have the `user.update.1234` permission and now I want to switch user 1234's group to the new group. I can't though, because I don't have the `group.update.MyNewGroup` permission. I also can't give that permission to myself because the whole point is to not allow people to give themselves permissions they don't already have.

TW80000 commented:
Groups could have `createdBy` support and we could use that as an alternative to checking that they have the appropriate `group.update.groupId` permission when changing another user's group.

Commit:
Add copyright header

Commit:
Update documentation

Commit:
Fix bug preventing group creation

The old code would throw an exception whenever trying to create a group
because the initial find would always return null for a new group.

Commit:
When creating a group, gain permission to update that group

Commit:
Add tests to prevent privilege escalation

Commit:
Fix bug preventing users from adding other users to groups

When you create a group and want to add someone to it, you must have
permission to update that user and you must have the update permission
for the group that you're putting them in. The previous code checked the
group they were currently in, not the group we were putting them in.

Commit:
Simplify logic

TW80000 commented:
Here is the test code:

```Java
// Prevent privilege escalation tests
// Tests to ensure that users can't escalate their privileges

import foam.dao.DAO;
import foam.nanos.auth.User;
import foam.nanos.auth.Group;
import foam.nanos.auth.Permission;
import foam.util.Auth;

try {
  bareUserDAO = (DAO) x.get("bareUserDAO");
  userDAO = (DAO) x.get("userDAO");
  groupDAO = (DAO) x.get("groupDAO");

  // Create a test user we'll eventually assign the group of.
  bareUserDAO.where(foam.mlang.MLang.EQ(User.EMAIL, "authenticatedgroupdaotestassigneduser@example.com")).removeAll();
  assignedUser = new User();
  assignedUser.setEmail("authenticatedgroupdaotestassigneduser@example.com");
  assignedUser.setGroup("basicUser");
  assignedUser = bareUserDAO.put(assignedUser);

  // Create a test group
  groupDAO.where(foam.mlang.MLang.EQ(Group.ID, "AuthenticatedGroupDAOTestGroup")).removeAll();
  group = new Group();
  group.setId("AuthenticatedGroupDAOTestGroup");
  group.setParent("basicUser");
  updatePermission = new Permission();
  updatePermission.setId("group.update.AuthenticatedGroupDAOTestGroup");
  createPermission = new Permission();
  createPermission.setId("group.create");
  updateAssignedUserPermission = new Permission();
  updateAssignedUserPermission.setId("user.update." + assignedUser.getId());
  Permission[] permissions = {
    updatePermission,
    createPermission,
    updateAssignedUserPermission
  };
  group.setPermissions(permissions);
  group = groupDAO.put(group);

  // Create a test user to sudo to
  bareUserDAO.where(foam.mlang.MLang.EQ(User.EMAIL, "AuthenticatedGroupDAOTest@example.com")).removeAll();
  user = new User();
  user.setEmail("AuthenticatedGroupDAOTest@example.com");
  user.setGroup("AuthenticatedGroupDAOTestGroup");
  user = bareUserDAO.put(user);
  userContext = Auth.sudo(x, user);

  // Try to create a new group that has a permission you don't have
  groupDAO.where(foam.mlang.MLang.EQ(Group.ID, "MorePowerful")).removeAll();
  morePowerful = new Group();
  morePowerful.setId("MorePowerful");
  morePowerful.setParent("basicUser");
  morePowerfulUpdatePermission = new Permission();
  morePowerfulUpdatePermission.setId("group.update.*");
  Permission[] morePowerfulPermissions = {
    morePowerfulUpdatePermission
  };
  morePowerful.setPermissions(morePowerfulPermissions);
  try {
    morePowerful = groupDAO.put_(userContext, morePowerful);
    test(false, "Users cannot create a group containing a permission that isn't implied by one they already have.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("Permission Denied. You do not have the 'group.update.*' permission."), "Users cannot create a group containing a permission that isn't implied by one they already have.");
  }

  // Try to add a permission you don't have to a group by updating
  try {
    group = group.fclone();
    group.setPermissions(morePowerfulPermissions);
    group = groupDAO.put_(userContext, group);
    test(false, "Users cannot update a group containing a permission that isn't implied by one they already have.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("Permission Denied. You do not have the 'group.update.*' permission."), "Users cannot update a group containing a permission that isn't implied by one they already have.");
  }

  // Try to set the parent of a group to "admin".
  try {
    group = group.fclone();
    group.setParent("admin");
    group = groupDAO.put_(userContext, group);
    test(false, "Users cannot set the parent of a group to a group containing a permission that isn't implied by one they already have.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("Permission Denied. You do not have the 'group.update.*' permission."), "Users cannot set the parent of a group to a group containing a permission that isn't implied by one they already have.");
  }

  // Create a group with a permission I do have
  groupDAO.where(foam.mlang.MLang.EQ(Group.ID, "AuthenticatedGroupDAOTestMyNewGroup")).removeAll();
  myNewGroup = new Group();
  myNewGroup.setId("AuthenticatedGroupDAOTestMyNewGroup");
  myNewGroup.setParent("basicUser");
  Permission[] myNewGroupPermissions = {
    createPermission
  };
  myNewGroup.setPermissions(myNewGroupPermissions);
  myNewGroup = groupDAO.put_(userContext, myNewGroup);
  test(myNewGroup != null, "Users can create new groups with permissions they have.");

  // Assign another user to a group I created.
  assignedUser = (User) assignedUser.fclone();
  assignedUser.setGroup("AuthenticatedGroupDAOTestMyNewGroup");
  try {
    assignedUser = userDAO.put_(userContext, assignedUser);
    test(assignedUser.getGroup().equals("AuthenticatedGroupDAOTestMyNewGroup"), "Can assign users to the new group.");
  } catch (Throwable t) {
    t.printStackTrace();
    test(false, "Can assign users to the new group.");
    print("INFO: " + t.getMessage());
  }

  // Try to change my own group.
  try {
    user = (User) user.fclone();
    user.setGroup("admin");
    user = userDAO.put_(userContext, user);
    test(false, "Users cannot set their own group.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("You cannot change your own group."), "Users cannot set their own group.");
  }
} catch (Throwable e) {
  e.printStackTrace();
  test(false, "An unexpected exception was thrown.");
}
```

TW80000 commented:
@kgrgreer @eatonk All comments addressed besides the one about where to put the new logic. Please give it another review when you can. Thank you.

kgrgreer commented on src/foam/nanos/auth/GiveUpdatePermissionDAO.java:
```diff
@@ -0,0 +1,51 @@
+    Group existing = (Group) getDelegate().find_(x, toPut.getId());
+    AuthService auth = (AuthService) x.get("auth");
+
+    if ( existing == null && ! auth.check(x, "group.update." + toPut.getId()) ) {
```
The way we addressed this in the past was by using wildcards for groups. For example, if we wanted some company admin to have the ability to create new groups we might give them the group.update.acme.* permission. Then they would create all of their groups as acme.operator, acme.admin, acme.accountant.level1, etc. This helps keep the group namespace orderly when being used by multiple companies. Also, the problem with the above approach is if I had an Acme.admin group who creates an acme.junioradmin group who creates acme.something else, then the original Acme.admin wouldn't have access (directly) to the new acme.something group, but by using hierarchical groups they would.

TW80000 replied:
Say we give some admin group for Acme the `"group.update.acme.*"` permission, but they create a group just called `"employee"` instead of `"acme.employee"` like we wanted them to. They still don't have permission to that group because they only have `"group.update.acme.*"`.

I guess we could solve that by making the `group.create` permission more fine-grained as well, something like `group.create.acme`. Then they would only be able to create groups prefixed with `"acme."`. Does that work with what you had in mind?

kgrgreer replied:
Yes, the system just wouldn't let them create groups with just the 'employee' id. The Auth decorator could even just imply the update permission if you had the create permission.

Commit:
Require another permission to change a user's group

Check for update permission on both old and new group when changing
another user's group.

Commit:
Fix mistake in logic

Commit:
Update test to account for updated requirement

Now we require update permission on both the old and new group to change
a user's group, so the test needed to be updated to reflect this change.

TW80000 commented:
Updated the code to now check for permissions on both the old and new group when changing another user's group. Also updated the test to reflect that.

The updated test code is below. The only difference is that the user we assign now starts in the group `PPETOriginalGroup` instead of `basicUser` and the user who's doing the assigning now has the `group.update.PPETOriginalGroup` permission.

```Java
// Prevent privilege escalation tests
// Tests to ensure that users can't escalate their privileges

import foam.dao.DAO;
import foam.nanos.auth.User;
import foam.nanos.auth.Group;
import foam.nanos.auth.Permission;
import foam.util.Auth;

try {
  bareUserDAO = (DAO) x.get("bareUserDAO");
  userDAO = (DAO) x.get("userDAO");
  groupDAO = (DAO) x.get("groupDAO");

  // Create a test user we'll eventually assign the group of.
  bareUserDAO.where(foam.mlang.MLang.EQ(User.EMAIL, "authenticatedgroupdaotestassigneduser@example.com")).removeAll();
  assignedUser = new User();
  assignedUser.setEmail("authenticatedgroupdaotestassigneduser@example.com");
  assignedUser.setGroup("PPETOriginalGroup");
  assignedUser = bareUserDAO.put(assignedUser);

  // Create a test group that our "weak" admin will be a part of.
  groupDAO.where(foam.mlang.MLang.EQ(Group.ID, "AuthenticatedGroupDAOTestGroup")).removeAll();
  group = new Group();
  group.setId("AuthenticatedGroupDAOTestGroup");
  group.setParent("basicUser");
  updateOriginalGroupPermission = new Permission();
  updateOriginalGroupPermission.setId("group.update.PPETOriginalGroup");
  updatePermission = new Permission();
  updatePermission.setId("group.update.AuthenticatedGroupDAOTestGroup");
  createPermission = new Permission();
  createPermission.setId("group.create");
  updateAssignedUserPermission = new Permission();
  updateAssignedUserPermission.setId("user.update." + assignedUser.getId());
  Permission[] permissions = {
    updateOriginalGroupPermission,
    updatePermission,
    createPermission,
    updateAssignedUserPermission
  };
  group.setPermissions(permissions);
  group = groupDAO.put(group);

  // Create a test user to sudo to
  bareUserDAO.where(foam.mlang.MLang.EQ(User.EMAIL, "AuthenticatedGroupDAOTest@example.com")).removeAll();
  user = new User();
  user.setEmail("AuthenticatedGroupDAOTest@example.com");
  user.setGroup("AuthenticatedGroupDAOTestGroup");
  user = bareUserDAO.put(user);
  userContext = Auth.sudo(x, user);

  // Try to create a new group that has a permission you don't have
  groupDAO.where(foam.mlang.MLang.EQ(Group.ID, "MorePowerful")).removeAll();
  morePowerful = new Group();
  morePowerful.setId("MorePowerful");
  morePowerful.setParent("basicUser");
  morePowerfulUpdatePermission = new Permission();
  morePowerfulUpdatePermission.setId("group.update.*");
  Permission[] morePowerfulPermissions = {
    morePowerfulUpdatePermission
  };
  morePowerful.setPermissions(morePowerfulPermissions);
  try {
    morePowerful = groupDAO.put_(userContext, morePowerful);
    test(false, "Users cannot create a group containing a permission that isn't implied by one they already have.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("Permission Denied. You do not have the 'group.update.*' permission."), "Users cannot create a group containing a permission that isn't implied by one they already have.");
  }

  // Try to add a permission you don't have to a group by updating
  try {
    group = group.fclone();
    group.setPermissions(morePowerfulPermissions);
    group = groupDAO.put_(userContext, group);
    test(false, "Users cannot update a group containing a permission that isn't implied by one they already have.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("Permission Denied. You do not have the 'group.update.*' permission."), "Users cannot update a group containing a permission that isn't implied by one they already have.");
  }

  // Try to set the parent of a group to "admin".
  try {
    group = group.fclone();
    group.setParent("admin");
    group = groupDAO.put_(userContext, group);
    test(false, "Users cannot set the parent of a group to a group containing a permission that isn't implied by one they already have.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("Permission Denied. You do not have the 'group.update.*' permission."), "Users cannot set the parent of a group to a group containing a permission that isn't implied by one they already have.");
  }

  // Create a group with a permission I do have
  groupDAO.where(foam.mlang.MLang.EQ(Group.ID, "AuthenticatedGroupDAOTestMyNewGroup")).removeAll();
  myNewGroup = new Group();
  myNewGroup.setId("AuthenticatedGroupDAOTestMyNewGroup");
  myNewGroup.setParent("basicUser");
  Permission[] myNewGroupPermissions = {
    createPermission
  };
  myNewGroup.setPermissions(myNewGroupPermissions);
  myNewGroup = groupDAO.put_(userContext, myNewGroup);
  test(myNewGroup != null, "Users can create new groups with permissions they have.");

  // Assign another user to a group I created.
  assignedUser = (User) assignedUser.fclone();
  assignedUser.setGroup("AuthenticatedGroupDAOTestMyNewGroup");
  try {
    assignedUser = userDAO.put_(userContext, assignedUser);
    test(assignedUser.getGroup().equals("AuthenticatedGroupDAOTestMyNewGroup"), "Can assign users to the new group.");
  } catch (Throwable t) {
    t.printStackTrace();
    test(false, "Can assign users to the new group.");
    print("INFO: " + t.getMessage());
  }

  // Try to change my own group.
  try {
    user = (User) user.fclone();
    user.setGroup("admin");
    user = userDAO.put_(userContext, user);
    test(false, "Users cannot set their own group.");
    print("INFO: No error was thrown.");
  } catch (foam.nanos.auth.AuthorizationException e) {
    test(e.getMessage().equals("You cannot change your own group."), "Users cannot set their own group.");
  }
} catch (Throwable e) {
  e.printStackTrace();
  test(false, "An unexpected exception was thrown.");
}
```
